src/System.Net.Security/src/System/Net/NTAuthentication.cs

```diff
@@ -0,0 +1,677 @@
+        {
+            Initialize(isServer, package, credential, spn, requestedContextFlags, channelBinding);
+        }
+
+        private class InitializeCallbackContext
+        {
+            internal InitializeCallbackContext(NTAuthentication thisPtr, bool isServer, string package, NetworkCredential credential, string spn, Interop.SspiCli.ContextFlags requestedContextFlags, ChannelBinding channelBinding)
+            {
+                this.thisPtr = thisPtr;
+                this.isServer = isServer;
+                this.package = package;
+                this.credential = credential;
+                this.spn = spn;
+                this.requestedContextFlags = requestedContextFlags;
+                this.channelBinding = channelBinding;
+            }
+
+            internal readonly NTAuthentication thisPtr;
+            internal readonly bool isServer;
+            internal readonly string package;
+            internal readonly NetworkCredential credential;
+            internal readonly string spn;
+            internal readonly Interop.SspiCli.ContextFlags requestedContextFlags;
+            internal readonly ChannelBinding channelBinding;
+        }
+
+        private static void InitializeCallback(object state)
+        {
+            InitializeCallbackContext context = (InitializeCallbackContext)state;
+            context.thisPtr.Initialize(context.isServer, context.package, context.credential, context.spn, context.requestedContextFlags, context.channelBinding);
+        }
+
+        private void Initialize(bool isServer, string package, NetworkCredential credential, string spn, Interop.SspiCli.ContextFlags requestedContextFlags, ChannelBinding channelBinding)
+        {
+            GlobalLog.Print("NTAuthentication#" + LoggingHash.HashString(this) + "::.ctor() package:" + LoggingHash.ObjectToString(package) + " spn:" + LoggingHash.ObjectToString(spn) + " flags :" + requestedContextFlags.ToString());
+            _tokenSize = SSPIWrapper.GetVerifyPackageInfo(GlobalSSPI.SSPIAuth, package, true).MaxToken;
+            _isServer = isServer;
+            _spn = spn;
+            _securityContext = null;
+            _requestedContextFlags = requestedContextFlags;
+            _package = package;
+            _channelBinding = channelBinding;
+
+            GlobalLog.Print("Peer SPN-> '" + _spn + "'");
+
+            //
+            // Check if we're using DefaultCredentials.
+            //
+
+            // TODO: This is a hack to make DefaultCredentials authentication work.
+            // Old code was checking for an internal type that is now internal to System.Net.Primitives: 
+            //      if (credential is SystemNetworkCredential)
+            if (credential.UserName == string.Empty && 
+                credential.Password == string.Empty &&
+                credential.Domain == string.Empty)
```

Comment:
@davidsh idea works. I've added a test to validate that checking against `CredentialCache.DefaultCredentials` isn't the same thing as receiving a `NetworkCredential` object with empty user, password and domain.